x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/cluster/metadata/MetadataMigrateToDataTiersRoutingService.java

```diff
@@ -0,0 +1,309 @@
+     * based on the provided node attribute name (we look at include, exclude and require rules) *ALL* the rules in the allocate action
+     * will be removed. All the rules are removed in order to allow for ILM to inject the {@link MigrateAction}.
+     * So for eg. this action:
+     *      allocate {
+     *          number_of_replicas: 0,
+     *          require: {data: warm},
+     *          include: {rack: one}
+     *      }
+     *  will become
+     *      allocate {
+     *          number_of_replicas: 0
+     *      }
+     *  Note that if the `allocate` action doesn't define any `number_of_replicas` it will be removed completely from the migrated policy.
+     *  As part of migrating the ILM policies we also update the cached phase definition for the managed indices to reflect the migrated
+     *  policy phase.
+     *
+     *  - loop through all the indices convert the index.routing.allocation.require.{nodeAttrName} or
+     *  index.routing.allocation.include.{nodeAttrName} setting (if present) to the corresponding data tier `_tier_preference` routing.
+     *  We are only able to convert the `frozen`, `cold`, `warm`, or `hot` setting values to the `_tier_preference`. If other
+     *  configuration values are present eg ("the_warm_nodes") the index will not be migrated.
+     *
+     * If no @param nodeAttrName is provided "data" will be used.
+     * If no @param indexTemplateToDelete is provided, no index templates will be deleted.
+     *
+     * This returns a new {@link ClusterState} representing the migrated state that is ready to use data tiers for index and
+     * ILM routing allocations. It also returns a summary of the affected abstractions encapsulated in {@link MigratedEntities}
+     */
+    public static Tuple<ClusterState, MigratedEntities> migrateToDataTiersRouting(ClusterState currentState,
+                                                                                  @Nullable String nodeAttrName,
+                                                                                  @Nullable String indexTemplateToDelete,
+                                                                                  NamedXContentRegistry xContentRegistry, Client client) {
+        IndexLifecycleMetadata currentMetadata = currentState.metadata().custom(IndexLifecycleMetadata.TYPE);
+        if (currentMetadata != null && currentMetadata.getOperationMode() != STOPPED) {
+            throw new IllegalStateException("stop ILM before migrating to data tiers, current state is [" +
+                currentMetadata.getOperationMode() + "]");
+        }
+
+        Metadata.Builder mb = Metadata.builder(currentState.metadata());
+        String removedIndexTemplateName = null;
+        if (Strings.hasText(indexTemplateToDelete) &&
+            currentState.metadata().getTemplates().containsKey(indexTemplateToDelete)) {
```

Comment:
I think it'd be nice to have a (debug) log message if a template was passed in, but not actually present in the cluster state and therefore couldn't be deleted